File: Sources/RedlineCore/FindingsDismissal.swift
// Which findings the user has already dealt with, and until when.
//
// A finding restates a habit that is still in the transcripts, so left alone it sits on screen
// for weeks after it has been read. Dismissing one hides it for a while; if it is still true
// when the snooze expires it comes back, because that is the only honest way to stop nagging
// without quietly dropping something real.
import Foundation

public struct FindingsDismissals: Codable, Equatable {
    /// Finding id to the moment it was dismissed. Ids are stable strings such as
    /// `mcp-unused`, so a dismissal survives a rescan.
    public private(set) var dismissed: [String: Date]

    public init(dismissed: [String: Date] = [:]) {
        self.dismissed = dismissed
    }

    public var isEmpty: Bool { dismissed.isEmpty }

    public func isHidden(_ id: String, snoozeDays: Int, now: Date = Date()) -> Bool {
        guard let at = dismissed[id] else { return false }
        // A snooze in the future means a clock that moved backwards, not a longer snooze
        guard at <= now else { return true }
        return now.timeIntervalSince(at) < Double(snoozeDays) * 86400
    }

    public mutating func dismiss(_ id: String, at: Date = Date()) {
        dismissed[id] = at
    }

    public mutating func restore(_ id: String) {
        dismissed[id] = nil
    }

    public mutating func restoreAll() {
        dismissed.removeAll()
    }

    /// Drops snoozes that have run out, so the file does not grow for every finding the
    /// checks have ever produced.
    public mutating func prune(snoozeDays: Int, now: Date = Date()) {
        dismissed = dismissed.filter { isHidden($0.key, snoozeDays: snoozeDays, now: now) }
    }
}

public enum FindingsDismissalStore {
    public static func url(home: URL? = nil) -> URL {
        (home ?? RedlineHome.url)
            .appendingPathComponent(".local/share/redline/findings-dismissed.json")
    }

    public static func load(from url: URL? = nil) -> FindingsDismissals {
        let url = url ?? FindingsDismissalStore.url()
        guard let data = try? Data(contentsOf: url) else { return FindingsDismissals() }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        // A file that exists but will not decode means dismissals silently came back.
        do {
            return try decoder.decode(FindingsDismissals.self, from: data)
        } catch {
            Diag.log.error("findings.dismissals_corrupt",
                           "dismissal file did not decode; starting over",
                           ["path": url.path, "error": String(describing: error)])
            return FindingsDismissals()
        }
    }

    @discardableResult
    public static func save(_ state: FindingsDismissals, to url: URL? = nil) -> Bool {
        let url = url ?? FindingsDismissalStore.url()
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(state) else {
            Diag.log.error("findings.dismissals_encode_failed", "could not encode dismissals")
            return false
        }
        do {
            try FileManager.default.createDirectory(at: url.deletingLastPathComponent(),
                                                    withIntermediateDirectories: true)
            try data.write(to: url, options: .atomic)
            try? FileManager.default.setAttributes([.posixPermissions: 0o600],
                                                   ofItemAtPath: url.path)
            return true
        } catch {
            Diag.log.error("findings.dismissals_save_failed", "could not write dismissals",
                           ["path": url.path, "error": String(describing: error)])
            return false
        }
    }
}

public extension FindingsReport {
    /// The same report with dismissed findings removed, and a count of what was hidden.
    ///
    /// Every surface reads its counts off this rather than off the raw report, so dismissing
    /// a finding quiets the menu bar line too. A panel that empties while the menu still
    /// claims four findings is worse than no dismissal at all.
    func visible(_ dismissals: FindingsDismissals, snoozeDays: Int,
                 now: Date = Date()) -> FindingsReport {
        let kept = findings.filter {
            !dismissals.isHidden($0.id, snoozeDays: snoozeDays, now: now)
        }
        return FindingsReport(generatedAt: generatedAt, windowDays: windowDays,
                              sessionsScanned: sessionsScanned, findings: kept,
                              hidden: findings.count - kept.count)
    }
}

File: windows/src/Redline.Core/FindingsDismissal.cs
// Which findings the user has already dealt with, and until when. A dismissal hides one for
// a while; if it is still true when the snooze runs out it comes back.
using System.Text.Json.Nodes;

namespace Redline.Core;

public sealed class FindingsDismissals : IEquatable<FindingsDismissals>
{
    private Dictionary<string, DateTimeOffset> _dismissed;

    /// Finding id to the moment it was dismissed. Ids are stable strings such as
    /// `mcp-unused`, so a dismissal survives a rescan.
    public IReadOnlyDictionary<string, DateTimeOffset> Dismissed => _dismissed;

    public FindingsDismissals(IDictionary<string, DateTimeOffset>? dismissed = null)
    {
        _dismissed = dismissed is null ? new() : new Dictionary<string, DateTimeOffset>(dismissed);
    }

    public bool IsEmpty => _dismissed.Count == 0;

    public bool IsHidden(string id, int snoozeDays, DateTimeOffset? now = null)
    {
        if (!_dismissed.TryGetValue(id, out var at)) return false;
        var t = now ?? DateTimeOffset.UtcNow;
        // A snooze in the future means a clock that moved backwards, not a longer snooze
        if (at > t) return true;
        return (t - at).TotalSeconds < snoozeDays * 86400.0;
    }

    public void Dismiss(string id, DateTimeOffset? at = null) => _dismissed[id] = at ?? DateTimeOffset.UtcNow;

    public void Restore(string id) => _dismissed.Remove(id);

    public void RestoreAll() => _dismissed.Clear();

    /// Drops snoozes that have run out, so the file does not grow for every finding ever.
    public void Prune(int snoozeDays, DateTimeOffset? now = null)
    {
        var t = now ?? DateTimeOffset.UtcNow;
        _dismissed = _dismissed.Where(kv => IsHidden(kv.Key, snoozeDays, t))
                               .ToDictionary(kv => kv.Key, kv => kv.Value);
    }

    public bool Equals(FindingsDismissals? o) => o is not null && _dismissed.Count == o._dismissed.Count &&
        _dismissed.All(kv => o._dismissed.TryGetValue(kv.Key, out var v) && v == kv.Value);
    public override bool Equals(object? obj) => Equals(obj as FindingsDismissals);
    public override int GetHashCode() => _dismissed.Count;
}

public static class FindingsDismissalStore
{
    public static string PathFor(string? home = null) =>
        Path.Combine(RedlineHome.DataDir(home), "findings-dismissed.json");

    public static FindingsDismissals Load(string? path = null)
    {
        path ??= PathFor();
        // A file that exists but will not decode means dismissals silently came back
        return StateFile.Load(path, "findings.dismissals_corrupt", "dismissal file did not decode; starting over",
            Decode, () => new FindingsDismissals());
    }

    public static bool Save(FindingsDismissals state, string? path = null)
    {
        path ??= PathFor();
        var dismissed = new JsonObject();
        foreach (var (id, at) in state.Dismissed.OrderBy(kv => kv.Key, StringComparer.Ordinal))
            dismissed[id] = StateFile.Stamp(at);
        return StateFile.Save(path, new JsonObject { ["dismissed"] = dismissed },
            "findings.dismissals_save_failed", "could not write dismissals");
    }

    private static FindingsDismissals Decode(JsonObject json)
    {
        if (json["dismissed"] is not JsonObject d) throw new FormatException("dismissed missing");
        var map = new Dictionary<string, DateTimeOffset>();
        foreach (var (id, node) in d) map[id] = StateFile.ParseDate(node);
        return new FindingsDismissals(map);
    }
}

public static class FindingsReportExt
{
    /// The same report with dismissed findings removed and a count of what was hidden. Every
    /// surface reads its counts off this, so dismissing quiets the menu line too.
    public static FindingsReport Visible(this FindingsReport report, FindingsDismissals dismissals,
                                         int snoozeDays, DateTimeOffset? now = null)
    {
        var t = now ?? DateTimeOffset.UtcNow;
        var kept = report.Findings.Where(f => !dismissals.IsHidden(f.Id, snoozeDays, t)).ToList();
        return new FindingsReport(report.GeneratedAt, report.WindowDays, report.SessionsScanned,
                                  kept, report.Findings.Count - kept.Count);
    }
}
